[clang/AST] Make it possible to use SwiftAttr in type context

xedin opened the pull request:
Swift ClangImporter now supports concurrency annotations on imported declarations and their parameters/results, to make it possible to use imported APIs in Swift safely there has to be a way to annotate individual parameters and result types with relevant attributes that indicate that e.g. a block is called on a particular actor or it accepts a `Sendable` parameter.

To faciliate that `SwiftAttr` is switched from `InheritableAttr` which is a declaration attribute to `DeclOrTypeAttr`. To support this attribute in type context we need access to its "Attribute" argument which requires `AttributedType` to be extended to include `Attr *` when available instead of just `attr::Kind` otherwise it won't be possible to determine what attribute should be imported.

xedin committed:
[clang/AST] Make it possible to use SwiftAttr in type context

Swift ClangImporter now supports concurrency annotations on imported declarations
and their parameters/results, to make it possible to use imported APIs in Swift
safely there has to be a way to annotate individual parameters and result types
with relevant attributes that indicate that e.g. a block is called on a particular
actor or it accepts a `Sendable` parameter.

To faciliate that `SwiftAttr` is switched from `InheritableAttr` which is a declaration
attribute to `DeclOrTypeAttr`. To support this attribute in type context we need access
to its "Attribute" argument which requires `AttributedType` to be extended to include
`Attr *` when available instead of just `attr::Kind` otherwise it won't be possible to
determine what attribute should be imported.

Xazax-hun commented on clang/lib/Sema/SemaType.cpp:
```diff
@@ -7163,7 +7217,8 @@ static QualType rebuildAttributedTypeWithoutNullability(ASTContext &Ctx,
   return Ctx.getAttributedType(Attributed->getAttrKind(), Modified,
-                               Attributed->getEquivalentType());
+                               Attributed->getEquivalentType(),
+															 Attributed->getAttr());
```
The formatting looks off here.

xedin replied:
✅

Xazax-hun commented on clang/lib/Sema/SemaType.cpp:
```diff
@@ -8732,6 +8786,11 @@ static void processTypeAttrs(TypeProcessingState &state, QualType &type,
       HandleHLSLParamModifierAttr(state, type, attr, state.getSema());
       attr.setUsedAsTypeAttr();
+			break;
+		}
```
The formatting might be a bit off here.

xedin replied:
Oh, it might be tabs vs. spaces :/

xedin replied:
✅

Xazax-hun commented on clang/lib/Sema/SemaType.cpp:
```diff
@@ -7147,6 +7147,60 @@ static bool HandleWebAssemblyFuncrefAttr(TypeProcessingState &State,
+
+  // If the attribute as attached to a paren move it closer to
+  // the declarator. This can happen in block declarations when
+  // an attribute is placed before `^` i.e. `(__attribute__((...)) ^)`.
```
I don't see a test case covering this scenario, should we add one?

xedin replied:
This actually applies more on the Swift side, I have multiple tests for this in ClangImporter but I'm not sure how to test this here since it doesn't really matter for clang AST where the attribute ends up being located.

xedin replied:
Done! I remembered why I didn't add that - it required `-fblocks`.

Xazax-hun reviewed (approved):
Overall looks good to me, I have one question and a couple nits inline. I think having access to Attr from AttributedType makes sense, I would not be surprised if this helped some other use cases like Lifetime annotations in Crubit. But please wait for a review from @erichkeane as well.

erichkeane reviewed (commented):
So I think this patch is doing too much, I'd like to see the changes to `AttributedType` split into a separate patch so we can talk about the `swift_attr` alone.

As far as moving this attribute to be a Type Attribute, that comes with a ton of concerns around templates.  I don't see anything set to instantiate teh arguments, so any value you give to it will not be instantiated (the attribute will as far as I can tell, be instantiated, but with no modifications to its arguments).

I would like to see a some tests (AST and not!) that validate the behavior of this on a template, and as a type being instantiated, to ensure it doesn't get lost in cases where you'd want it to stay.

Also, we need a release note.

hnrklssn commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -5125,24 +5127,54 @@ QualType ASTContext::getUnresolvedUsingType(
+                                       QualType equivalentType,
+                                       const Attr *attr) const {
   llvm::FoldingSetNodeID id;
   AttributedType::Profile(id, attrKind, modifiedType, equivalentType);
```
`AttributedType` is uniqued based on `attrKind`, but not `attr`. So calling this function may return a type with a pointer to a different `Attr` (although of the same kind). Is this a problem? Do we want to fully unique `AttributedType` based on it's `Attr*`?

The other option I believe is to fetch the `Attr*` from `AttributedTypeLoc` (which you can get through the `TypeSourceInfo` from `Decl` or `ExplicitCastExpr`) rather than the `AttributedType`, but doing that for generic expressions gets quite hairy.

xedin replied:
I think adding `attr` to profile is a better approach, I'll look into that.

xedin replied:
✅

xedin commented on clang/include/clang/AST/Type.h:
```diff
@@ -6037,13 +6038,24 @@ class AttributedType : public Type, public llvm::FoldingSetNode {
-        ModifiedType(modified), EquivalentType(equivalent) {
+        Attribute(attr), ModifiedType(modified),
+        EquivalentType(equivalent) {
     AttributedTypeBits.AttrKind = attrKind;
```
If I remember correct the `Attr *` is not always there (nullable or something) that's why we need both `Attr *` and `attrKind`

Xazax-hun replied:
I see, thanks! In that case it makes sense to me to encode this invariant in the constructor as an assertion (either `attr` is null, or it has the same kind as `attrKind`).

xedin replied:
Sure!

xedin replied:
In fact I added that assert to `getAttribute(...)` method on ASTContext already.

xedin replied:
✅

xedin commented:
I addressed most of the comments. A few points to add:

- I'm not sure what to make of the comment regarding templates if somebody could point me to the right direction I can make appropriate changes
- I'd need some guidance regarding release note
- I'd prefer to keep AttributedType changes in this PR since they are integral part of the swift_attr change.

Xazax-hun commented:
I think the interesting template cases are like things like:

```
template<typename T> struct S {
  T foo();
};

S<int> a;
S<SWIFT_SENDABLE int> b;

void f() {
  a.foo();
  b.foo();
}
```

The question is, whether we actually have the type attributes where we expect to have them, or do we lose them?
Hopefully @erichkeane can correct me if I'm wrong. 

That being said, one question is whether we actually need type attributes to be propagated in this case for Swift's interop. 

For the release notes, you'd want to mention that `swift_attr` can now be applied to types (maybe a small motivation why) in `clang/docs/ReleaseNotes.rst`. It has a section for attribute changes.

erichkeane commented:
> I think the interesting template cases are like:
> 
> ```
> template<typename T> struct S {
>   T foo();
> };
> 
> S<int> a;
> S<SWIFT_SENDABLE int> b;
> 
> void f() {
>   a.foo();
>   b.foo();
> }
> ```
> 
> The question is, whether we actually have the type attributes where we expect to have them, or do we lose them? Hopefully @erichkeane can correct me if I'm wrong.
> 
> That being said, one question is whether we actually need type attributes to be propagated in this case for Swift's interop.
> 
> For the release notes, you'd want to mention that `swift_attr` can now be applied to types (maybe a small motivation why) in `clang/docs/ReleaseNotes.rst`. It has a section for attribute changes.

You're correct, the problem with attributed type is that it is going to be lost A LOT every time we canonicalize the type, else they need to be different types, and thus those would be different instantiations.  This is SOMETIMES not problematic (and sometimes even wanted, see vector types), but it IS problematic if you want your attribute to not cause problems being used in containers/etc, since `std::is_same<int, MY_INT_ATTRIBUTED_TYPE>` being 'false' is problematic for many use cases.

xedin commented:
> That being said, one question is whether we actually need type attributes to be propagated in this case for Swift's interop.

I don't think we do, we need swift_attr to help us augment Objective-C APIs to make existing frameworks pray well with Swift concurrency. I think C++ Interoperability only uses them in limited set of cases which don't involve templates.

xedin commented:
Updated release notes and added block tests requested by @Xazax-hun.

xedin commented:
@erichkeane friendly ping about this, I replied to your last question wondering if it would be okay to land this or should I close since it has been awhile now.

erichkeane commented:
> > That being said, one question is whether we actually need type attributes to be propagated in this case for Swift's interop.
> 
> I don't think we do, we need swift_attr to help us augment Objective-C APIs to make existing frameworks play well with Swift concurrency. I think C++ Interoperability only uses them in limited set of cases which don't involve templates.

What about ObjC++?  We support that as well, and I'd like for us to not lose that ability too.

erichkeane commented on clang/include/clang/AST/TypeProperties.td:
```diff
@@ -668,12 +668,16 @@ let Class = AttributedType in {
+  def : Property<"attrKind", AttrKind> {
     let Read = [{ node->getAttrKind() }];
   }
+  def : Property<"attribute", Attr> {
```
Is there reason to store both `attribute` and `attrKind` here?  Should we just implement `attrKind` as `attribute.getKind` ?

xedin replied:
Yes, Attr is not always available (nullability related attributes only have a kind and don't allocate Attr).

erichkeane commented on clang/lib/AST/Type.cpp:
```diff
@@ -4104,6 +4105,19 @@ bool RecordType::hasConstFields() const {
+                               QualType equivalent)
+    : Type(Attributed, canon, equivalent->getDependence()), Attribute(attr),
+      ModifiedType(modified), EquivalentType(equivalent) {
+  AttributedTypeBits.AttrKind = attrKind;
```
I think it'd be lovely to get these bits back.  Perhaps just remove attrKind from this `ctor` and just implement the above one instead.

xedin replied:
I don't think we can do that, see https://github.com/llvm/llvm-project/pull/108631/files#r1823004595. @Xazax-hun had a comment to that effect already.

erichkeane reviewed (commented):
I think this would be a great opportunity to save a bit of memory on `AttributedType`, else this is probably ok to me.

xedin commented:
@erichkeane Do I need to re-run CI? I'd need some help with that since I don't have any permissions :)

erichkeane commented:
> @erichkeane Do I need to re-run CI? I'd need some help with that since I don't have any permissions :)

IDK how to do that other than push another 'small' patch with a merge commit or something.  You don't HAVE to, but there is of course increased risk of revert if you don't.  I might suggest just committing 1st thing in your morning, then dealing with the build bots during the day as they come up

xedin commented:
No worries, I can just rebase and push force, that should do it :)

xedin commented:
@erichkeane CI are green, could you please merge since I don't have permissions?...

xedin commented:
I think if `noderef` attribute instantiations in your example are considered unique when they appear in different spots then we should have two distinct attributed types but it sounds like both of `noderef` should have one `Attr *` instantiation...

bevin-hansson commented:
Well, as far as I know Attrs are created on the spot and aren't uniqued. Perhaps that should happen for these kinds of Attrs.
